cmd/stolonctl/cmd/clusterdata.go

```diff
@@ -15,32 +15,76 @@
 package cmd
 
 import (
+	"context"
 	"encoding/json"
-
+	"fmt"
 	cmdcommon "github.com/sorintlab/stolon/cmd"
+	"github.com/sorintlab/stolon/internal/cluster"
+	"github.com/sorintlab/stolon/internal/store"
+	"io"
+	"io/ioutil"
+	"os"
 
 	"github.com/spf13/cobra"
 )
 
 var cmdClusterData = &cobra.Command{
 	Use:   "clusterdata",
-	Run:   clusterdata,
-	Short: "Retrieve the current cluster data",
+	Short: "Manage current cluster data",
 }
 
-type clusterdataOptions struct {
+type clusterdataReadOptions struct {
 	pretty bool
 }
 
-var clusterdataOpts clusterdataOptions
+type Reader interface {
```

Comment:
I don't understand why this interface and the below implementation is needed. I think that the functions `readClusterData`, `readClusterDataFromFile` and `readClusterDataFromStdIn` don't really need it since they receive the filename from the global config `readClusterdataOpts`.

Or is this just needed for the tests to mock the reader? If so isn't there a way to do this without "bloating" the code?

Reply:
We introduced `Reader` interface only for **unit testing**. We added unit tests using the MockReader (which are generated using interface Reader). If you think, it is bloating the code, we can remove the interface and write integration tests. Thoughts?

We are using `readClusterDataFromFile`, `readClusterDataFromStdIn` to read the new cluster data to be written to the new store. We will read from file or stdin based on `writeClusterdataOpts.file`. So we need both the methods. Thoughts ?

Reply:
Hi @sgotti. We **removed** `Reader` interface and methods `readClusterData`, `readClusterDataFromFile` and `readClusterDataFromStdIn`. We are using different way to achieve the unit testing.

Now we sending reader to `writeClusterdata` based on user input. For files, we are sending `File`(as File is a Reader) and `Stdin` (as Stdin is a Reader). For testing we are sending simple `StringReader`. So we able **easily test without mocking** Reader.

**Thanks for your inputs. It helped a lot and we removed lot of bloating code**. Please have a look and provide your inputs.